### ISSUE TITLE
issues in running the demo aesccm.c

### ISSUE THREAD

**Issue by nemr-maher** [openssl/openssl#5245]
Hi i'm trying to run the aesccm.c file on windows 10 with OpenSSL 1.1.0e 

I tried the following : 

`gcc test.c -o test`

`gcc -I C:/OpenSSL-Win64 -o test -lcrypto test.c`

I keep getting the following error :

`test.c:49:30: error: 'EVP_CTRL_AEAD_SET_IVLEN' undeclared (first use in this function)
     EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_SET_IVLEN, sizeof(ccm_nonce),
                              ^~~~~~~~~~~~~~~~~~~~~~~
test.c:49:30: note: each undeclared identifier is reported only once for each function it appears in
test.c:52:30: error: 'EVP_CTRL_AEAD_SET_TAG' undeclared (first use in this function)
     EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_SET_TAG, sizeof(ccm_tag), NULL);
                              ^~~~~~~~~~~~~~~~~~~~~
test.c:67:30: error: 'EVP_CTRL_AEAD_GET_TAG' undeclared (first use in this function)
     EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_GET_TAG, 16, outbuf);
                              ^~~~~~~~~~~~~~~~~~~~~
test.c: In function 'aes_ccm_decrypt':
test.c:86:30: error: 'EVP_CTRL_AEAD_SET_IVLEN' undeclared (first use in this function)
     EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_SET_IVLEN, sizeof(ccm_nonce),
                              ^~~~~~~~~~~~~~~~~~~~~~~
test.c:89:30: error: 'EVP_CTRL_AEAD_SET_TAG' undeclared (first use in this function)
     EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_SET_TAG,`

**Comment by mattcaswell** [comment]
Are you running this from an MS command prompt or MSYS2, or something else? Where did you get your gcc from?

**Comment by bluetiger30** [comment]
I'm using MS command prompt in administrator mode .

GCC is from cygwin.

bellow are the gcc version details :

```
gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-pc-cygwin/6.4.0/lto-wrapper.exe
Target: x86_64-pc-cygwin
Configured with: /cygdrive/i/szsz/tmpp/gcc/gcc-6.4.0-4.x86_64/src/gcc-6.4.0/configure --srcdir=/cygdrive/i/szsz/tmpp/gcc/gcc-6.4.0-4.x86_64/src/gcc-6.4.0 --prefix=/usr --exec-prefix=/usr --localstatedir=/var --sysconfdir=/etc --docdir=/usr/share/doc/gcc --htmldir=/usr/share/doc/gcc/html -C --build=x86_64-pc-cygwin --host=x86_64-pc-cygwin --target=x86_64-pc-cygwin --without-libiconv-prefix --without-libintl-prefix --libexecdir=/usr/lib --enable-shared --enable-shared-libgcc --enable-static --enable-version-specific-runtime-libs --enable-bootstrap --enable-__cxa_atexit --with-dwarf2 --with-tune=generic --enable-languages=ada,c,c++,fortran,lto,objc,obj-c++ --enable-graphite --enable-threads=posix --enable-libatomic --enable-libcilkrts --enable-libgomp --enable-libitm --enable-libquadmath --enable-libquadmath-support --enable-libssp --enable-libada --disable-symvers --with-gnu-ld --with-gnu-as --with-cloog-include=/usr/include/cloog-isl --without-libiconv-prefix --without-libintl-prefix --with-system-zlib --enable-linker-build-id --with-default-libstdcxx-abi=gcc4-compatible
Thread model: posix
gcc version 6.4.0 (GCC)
```

**Comment by nemr-maher** [comment]
I'm using MS command prompt in administrator mode .

GCC is from cygwin.

bellow are the gcc version details :

```
gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-pc-cygwin/6.4.0/lto-wrapper.exe
Target: x86_64-pc-cygwin
Configured with: /cygdrive/i/szsz/tmpp/gcc/gcc-6.4.0-4.x86_64/src/gcc-6.4.0/configure --srcdir=/cygdrive/i/szsz/tmpp/gcc/gcc-6.4.0-4.x86_64/src/gcc-6.4.0 --prefix=/usr --exec-prefix=/usr --localstatedir=/var --sysconfdir=/etc --docdir=/usr/share/doc/gcc --htmldir=/usr/share/doc/gcc/html -C --build=x86_64-pc-cygwin --host=x86_64-pc-cygwin --target=x86_64-pc-cygwin --without-libiconv-prefix --without-libintl-prefix --libexecdir=/usr/lib --enable-shared --enable-shared-libgcc --enable-static --enable-version-specific-runtime-libs --enable-bootstrap --enable-__cxa_atexit --with-dwarf2 --with-tune=generic --enable-languages=ada,c,c++,fortran,lto,objc,obj-c++ --enable-graphite --enable-threads=posix --enable-libatomic --enable-libcilkrts --enable-libgomp --enable-libitm --enable-libquadmath --enable-libquadmath-support --enable-libssp --enable-libada --disable-symvers --with-gnu-ld --with-gnu-as --with-cloog-include=/usr/include/cloog-isl --without-libiconv-prefix --without-libintl-prefix --with-system-zlib --enable-linker-build-id --with-default-libstdcxx-abi=gcc4-compatible
Thread model: posix
gcc version 6.4.0 (GCC)
```

**Comment by p-steuer** [comment]
Looks to me like gcc includes an older evp.h, which only defines  EVP_CTRL_CCM_* and not  EVP_CTRL_AEAD_* .

**Comment by nemr-maher** [comment]
how can I update that ?

**Comment by p-steuer** [comment]
Did you try 
gcc -I C:/OpenSSL-Win64/include/ -o test -lcrypto test.c

?

(i assume your 1.1.0 source is OpenSSL-Win64)

**Comment by mattcaswell** [comment]
Don't mix cygwin and ms command prompt environments! The cygwin gcc is special and targeted for the cygwin shell. As @p-steuer suggests my guess is that it is picking up the includes from a different version of OpenSSL. Where did you get your copy of OpenSSL from? Did you download it as a binary or compile it from source?

**Comment by nemr-maher** [comment]
same result .
I found an evp.h in the following directory (C:\cygwin64\usr\include\openssl) and it only had  EVP_CTRL_AEAD_TLS1_AAD and EVP_CTRL_AEAD_SET_MAC_KEY. 

I checked the cygwin installer it shows the following :
![image](https://user-images.githubusercontent.com/30926789/35730536-6a6642e6-080a-11e8-9542-389e0bc8e68d.png)

so looks like cygwin is using 1.0.2 , how can I update that to the current one ?

**Comment by nemr-maher** [comment]
I got the openssl.exe from [here](https://slproweb.com/products/Win32OpenSSL.html)

**Comment by mattcaswell** [comment]
Right. That is a native windows version of OpenSSL not a cygwin version. Are you wanting to target native Windows or Cygwin?

**Comment by nemr-maher** [comment]
Cygwin

**Comment by mattcaswell** [comment]
Ok, so don't use the slpro OpenSSL version you downloaded. You need to use a version specifically compiled for cygwin - and use the cygwin shell for compiling etc. You have a few options:

1) Stick with OpenSSL 1.0.2. This is still supported but there are some differences in the API between 1.0.2 and the latest 1.1.0

2) Get a new version from cygwin. I don't know if such an update exists or not. You'd have to ask the cygwin guys.

3) Compile 1.1.0 from source. In which case download the 1.1.0 source from https://www.openssl.org/source/ and follow the build instructions in the README.

**Comment by nemr-maher** [comment]
Thanks , I'm interested in the AES-CCM mainly so it would have to be 1.1.0  I think as 1.0.2 doesn't support it and I got the latest cygwin version.

I tried to compile the latest Openssl version , and had a look at the README , I get stuck at the nmake command as it indicates `-bash: nmak: command not found` how can I get that ?
Thanks for your time an I appreciate your help

**Comment by mattcaswell** [comment]
1.0.2 does support CCM. Here are the 1.0.2 docs on the topic:

https://www.openssl.org/docs/man1.0.2/crypto/EVP_EncryptInit.html

As for compiling - the thing you have to realise is that Cygwin is, to all intents and purposes, a UNIX environment. So actually you need to use UNIX style "make" not "nmake" (which is for native windows). I suggest you start by reading the cygwin instructions in NOTES.WIN here:

https://github.com/openssl/openssl/blob/OpenSSL_1_1_0-stable/NOTES.WIN#L58

**Comment by nemr-maher** [comment]
Thanks again for your help.

I decided to go for the native windows , managed to build the current version of OpenSSL 1.1.1 and added it to Path. I'm using the `x64 Native Tools Command Prompt for VS 2017` to compile the `aesccm.c` file using the follwoing command `cl aesccm.c`
I get the follwoing error : 

![image](https://user-images.githubusercontent.com/30926789/35745412-1f69eaaa-083b-11e8-9332-d5cdad405498.png)
and this is the image of lib directory 
![image](https://user-images.githubusercontent.com/30926789/35745459-430e9d98-083b-11e8-9bc3-ccb4dcbfa371.png)

**Comment by mattcaswell** [comment]
I'm not really an expert on the VC++ tool chain. However presumably you need to tell `cl` where the include files live

**Comment by nemr-maher** [comment]
Thanks , I sorted all the build issues out , however now the programme doesn't print anything except the 1st 2 statements :
![image](https://user-images.githubusercontent.com/30926789/35813533-c458b1c0-0a8b-11e8-8c15-e76b4579d300.png)

**Comment by mattcaswell** [comment]
See:

https://www.openssl.org/docs/faq.html#PROG3

Compile your application with `/MD`.

**Comment by nemr-maher** [comment]
I have already done that : 
![image](https://user-images.githubusercontent.com/30926789/35814457-0805e774-0a8e-11e8-8454-92d90c018c24.png)

**Comment by levitte** [comment]
Did you read the *whole* FAQ entry?

**Comment by mattcaswell** [comment]
Did you link your application with applink.c (as mentioned in the FAQ)?

**Comment by nemr-maher** [comment]
Thanks got it working now , I had the applink but I wasn't running VS as admin which was preventing it from creating the library. Thanks again for you help and support . I do apologies for any inconvenience
